### PR TITLE
Added exceptSelect function for Query

### PR DESCRIPTION
**Pull request by ghost**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by samdark** [comment]
I've got it. The case is a table with 100s of columns and the need to select everything except "status", "is_deleted".

**Comment by samdark** [comment]
Personally I both like and don't like this addition. I've been in situation like that and wished there was a method like that but that was only once in like 5 years.

**Comment by samdark** [comment]
@yiisoft/core-developers, @yiisoft/reviewers opinions?

**Comment by klimov-paul** [comment]
I do not consider use case to be common enough. If it would SQL native syntax suported it long ago.
Having a database table with so much columns, so you need `exceptSelect` feature, usually indicates a poor database structure design.
Also such functionality can be easily composed into an `ActiveQuery` behavior or trait and be applied at will. Also building a list of fields can be moved into a separated helper.

**Comment by Inocence** [comment]
I thought about you said, if I build a list , i must update it when i changed sql structure,then, every table just have 20 columns,but i meet this problem in many projects,every time i must write same code to resolve it

File: framework/db/Query.php [hunk 1]
```diff
@@ -597,6 +597,35 @@ public function addSelect($columns)
...
+        $tableSchema = Yii::$app->db->getSchema()->getTableSchema($realTable);
+        if ($tableSchema === null) {
+            throw new InvalidConfigException('The table does not exist: ' . $realTable);
+        }
+        $col = $tableSchema->getColumnNames();
+        $exceptColumns = is_array($exceptColumns) ? $exceptColumns : preg_split('/\s*,\s*/', trim($exceptColumns), -1, PREG_SPLIT_NO_EMPTY);
+        foreach($exceptColumns as $v) {
+            unset($col[array_keys($col, $v)[0]]);
+        }
+        $this->select = array_merge($this->select, array_map(function($v) use($asTable) {return $asTable.'.'.$v;}, $col));
```
Comment (bscheshirwork) [inline framework/db/Query.php]: $v is tooo short (codeclimate min length is 3) and not relate to `column name`

File: framework/db/Query.php [hunk 2]
```diff
@@ -597,6 +597,35 @@ public function addSelect($columns)
...
+     */
+    public function exceptSelect($asTable, $realTable, $exceptColumns)
+    {
+        $tableSchema = Yii::$app->db->getSchema()->getTableSchema($realTable);
+        if ($tableSchema === null) {
+            throw new InvalidConfigException('The table does not exist: ' . $realTable);
+        }
+        $col = $tableSchema->getColumnNames();
+        $exceptColumns = is_array($exceptColumns) ? $exceptColumns : preg_split('/\s*,\s*/', trim($exceptColumns), -1, PREG_SPLIT_NO_EMPTY);
+        foreach($exceptColumns as $v) {
```
Comment (bscheshirwork) [inline framework/db/Query.php]: `$v` is too short (codeclimate), we can use `$value` instead

**Comment by Inocence** [comment]
And,maybe it is not very common,but I think it is necessary. Anyway I really hope you can accept this feature ,so I can never write same code again,thanks

**Comment by Inocence** [comment]
Ok,thank you to review the code,i will rename variables tomorrow

File: framework/db/Query.php (hunk 2 above)
Comment (Inocence) [reply framework/db/Query.php]: Ok,thank you to review the code,i will rename variables tomorrow

**Comment by bscheshirwork** [comment]
@Inocence 
> I thought about you said, if I build a list , i must update it when i changed sql structure,then, every table just have 20 columns,but i meet this problem in many projects,every time i must write same code to resolve it

> And,maybe it is not very common,but I think it is necessary. Anyway I really hope you can accept this feature ,so I can never write same code again,thanks

I think your can use your own repo with this behaviour just now and use it in all projects. 
For example: I have repo [bscheshirwork/yii2-cubs](https://github.com/bscheshirwork/yii2-cubs)

**Comment by Inocence** [comment]
Ok, I got it,thanks.

**Comment by cebe** [comment]
We currently have `addSelect()` to add columns to the select clause, maybe we can have a `removeSelect()` that works complementary to that.

**Comment by bscheshirwork** [comment]
https://github.com/yiisoft/yii2/blob/376006a2d5048764ebad70a3fd1077ceb4989352/framework/db/Query.php#L568-L597

Have only one parameter `$columns` and independent of table structure or database driver.  
Revere of this will expect same signature. 

i.e. something like 
```php
public function removeSelect($columns, ?array $alias = null)
```

Can I will use `removeSelect` for query of `\yii\redis\ActiveRecord` model?

**Comment by bscheshirwork** [comment]
Additional expects 
```php
$query
->addSelect(["*", "CONCAT(first_name, ' ', last_name) AS full_name"])
->removeSelect(["*", "CONCAT(first_name, ' ', last_name) AS full_name"])
```
still `$query` is unchanged

**Comment by Inocence** [comment]
i think i should change the function to make it works like `addSelect()`

**Comment by samdark** [comment]
I like what @cebe proposed.

**Comment by samdark** [comment]
Won't make it into 2.0.

**Pull request closed without merging**